File: The_Algorithm_Design_Manual/4_Sorting_and_Searching/HeapSort.java
public class HeapSort {
    public void sort(int[] arr) {
        heapify(arr);

        for(int len=arr.length; len > 1; ) {
            int max = arr[0];
            arr[0] = arr[len-1];
            arr[len-1] = max;
            siftDown(arr, 0, --len);
        }
    }

    private void heapify(int[] arr) {
        for (int i=(arr.length-2)/2; i>=0; i--) {
            siftDown(arr, i, arr.length);
        }
    }

    private void siftDown(int[] arr, int start, int end) {
        int root = start;

        while (2*root+1 < end) {
            int left = 2*root + 1;
            int right = 2*root + 2;
            int sift = left;
            if (right < end && arr[left] < arr[right]) {
                sift = right;
            }

            if (arr[root] < arr[sift]) {
                int tmp = arr[sift];
                arr[sift] = arr[root];
                arr[root] = tmp;
                root = sift;
            } else {
                break;
            }
        }
    }

    public static void main(String[] args) {
        int[] arr = new int[] {2, 3, 5, 4, 1, 10, 9};

        HeapSort hs = new HeapSort();
        hs.sort(arr);

        for(int i=0; i<arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}

File: The_Algorithm_Design_Manual/4_Sorting_and_Searching/heap_sort.c
/**************************************************************/
/* File: heap_sort.c                                          */
/* Heap sort.                                                 */
/**************************************************************/

#include <stdio.h>

void siftDown(int* arr, int size, int i) {
    int left = 2*i + 1;
    int right = 2*i + 2;

    if (left < size) { // it is not the leaf node
        int sift = left;

        if (right < size && arr[left] < arr[right]) {
            sift = right;
        }

        if (arr[i] < arr[sift]) {
            int tmp = arr[i];
            arr[i] = arr[sift];
            arr[sift] = tmp;
            siftDown(arr, size, sift);
        }
    }
}

void heapify(int* arr, int size) {
    int i;
    for (i=(size-2)/2;i>=0;i--) {
        siftDown(arr, size, i);
    }
}

void heapSort(int* arr, int size) {
    heapify(arr, size);

    int n = size;
    while (n > 1) {
        int max = arr[0];
        arr[0] = arr[n-1];
        arr[n-1] = max;
        n--;
        siftDown(arr,n,0);
    }
}

int main(int argc, char* argv[]) {
    #define SIZE 7
    int arr[SIZE] = {2,3,1,10,8,7,9};
    heapSort(arr, SIZE);

    int i;
    for(i=0; i<SIZE; i++) {
        printf("%d ", arr[i]);
    }

    printf("\n");

    return 0;
}
